torch/utils/data/datapipes/_typing.py

```diff
@@ -421,6 +421,7 @@ def _set_datapipe_valid_iterator_id(datapipe):
             datapipe._valid_iterator_id = 0
         else:
             datapipe._valid_iterator_id += 1
+        datapipe.reset()
```

Comment:
@ejguan  `reset()` has been added here to make sure `reset()` is called every time when `__iter__` is called and updates the `_valid_iterator_id`